File: motor_de_jogo.py
# domino_v3.py
import random
from collections import deque
import csv
import os
import uuid
import time
from typing import List, Dict, Optional
from core.peca import Peca
from core.jogador import Jogador
from core.tabuleiro import Tabuleiro
from core.dupla import Dupla
from utilidades.distribuicao import distribuir_jogadores
from regras.game_logic import (
    proximo_jogador_obj,
    determinar_tipo_batida,
    pontuacao_por_tipo,
    determinar_vencedor_travamento,
)

def simular_rodada(
    jogadores: List[Jogador],
    jogador_inicial_nome: Optional[str] = None,
    *,
    duplas: dict[str, Dupla] | None = None,
    pontos_para_vencer: int = 6,
):
    tabuleiro = Tabuleiro()
    historico = []
    estados = []
    ordem_jogada = 1
    passes_por_jogador = {j.nome: 0 for j in jogadores}

    # Identificar o jogador com o maior duplo (se não especificado)
    if jogador_inicial_nome is None:
        maior_duplo = -1
        jogador_inicial = None
        for j in jogadores:
            for p in j.mao:
                if p.is_duplo() and p.lado1 > maior_duplo:
                    maior_duplo = p.lado1
                    jogador_inicial = j
        if not jogador_inicial:
            return {"erro": "Nenhum duplo encontrado"}
    else:
        jogador_inicial = next(j for j in jogadores if j.nome == jogador_inicial_nome)

    jogador_atual = jogador_inicial
    motivo_fim = None
    tipo_batida = None
    vencedor_rodada = None
    pontuacao_rodada = 0

    while True:
        jogadas = jogador_atual.jogadas_validas(tabuleiro.obter_pontas())

        if jogadas:
            if ordem_jogada == 1 and jogador_inicial_nome is None:
                # Primeira jogada: deve ser o maior duplo
                peca_jogada = next(p for p in jogador_atual.mao if p.is_duplo() and p.lado1 == maior_duplo)
            else:
                peca_jogada = jogador_atual.escolher_peca(
                    tabuleiro,
                    jogadores,
                    duplas=duplas or {},
                    passes_jog=passes_por_jogador,
                    pontos_para_vencer=pontos_para_vencer,
                )

            jogador_atual.remover_peca(peca_jogada)
            lado = tabuleiro.jogar(peca_jogada)
            tipo = "batida" if not jogador_atual.mao else "jogada"

            historico.append({
                "ordem": ordem_jogada,
                "jogador": jogador_atual.nome,
                "tipo": tipo,
                "peca": (peca_jogada.lado1, peca_jogada.lado2),
                "lado": lado
            })

            tabuleiro.resetar_passes()

            if not jogador_atual.mao:
                tipo_batida = determinar_tipo_batida(peca_jogada, tabuleiro.obter_pontas())
                pontuacao_rodada = pontuacao_por_tipo(tipo_batida)
                vencedor_rodada = jogador_atual.nome
                motivo_fim = "batida"

        else:
            historico.append({
                "ordem": ordem_jogada,
                "jogador": jogador_atual.nome,
                "tipo": "passe"
            })
            passes_por_jogador[jogador_atual.nome] += 1
            jogador_atual.registrar_passe(tabuleiro.obter_pontas())
            tabuleiro.registrar_passe()

            if tabuleiro.passes_consecutivos == 4:
                vencedor_rodada, pontuacao_rodada = determinar_vencedor_travamento(jogadores)
                motivo_fim = "travamento"
                tipo_batida = "travamento"

        estados.append({
            "ordem_jogada": ordem_jogada,
            "jogador": jogador_atual.nome,
            "tipo": historico[-1]["tipo"],
            "peca": historico[-1].get("peca"),
            "lado": historico[-1].get("lado"),
            "tabuleiro": [(p.lado1, p.lado2) for p in tabuleiro.pecas],
            "maos": {j.nome: [(p.lado1, p.lado2) for p in j.mao] for j in jogadores},
            "tipo_batida": tipo_batida,
            "motivo_fim": motivo_fim,
            "vencedor_rodada": vencedor_rodada,
            "jogadas": historico,
            "jogadas_disponiveis": [(p.lado1, p.lado2) for p in jogadas]
        })

        if motivo_fim:
            break

        ordem_jogada += 1
        jogador_atual = proximo_jogador_obj(jogadores, jogador_atual)

    return {
        "estados": estados,
        "final": {
            "tipo_batida": tipo_batida,
            "motivo_fim": motivo_fim,
            "vencedor_rodada": vencedor_rodada,
            "pontuacao_rodada": pontuacao_rodada
        }
    }


def simular_partida(
    pontos_para_vencer: int = 6,
    pontuacao_por_jogador: Optional[dict] = None,
    estrategias: Optional[dict] = None,
) -> dict:
    """Simula uma partida completa.

    Parameters
    ----------
    pontos_para_vencer: int
        Pontuação alvo para que uma dupla vença a partida.
    pontuacao_por_jogador: dict | None
        Dicionário opcional para acumular a pontuação ao longo de múltiplas
        partidas.
    estrategias: dict | None
        Mapeamento ``nome_jogador -> estratégia`` a ser utilizado na
        distribuição das peças. Valores podem ser callables, objetos com
        ``escolher_peca`` ou subclasses de :class:`Jogador`.
    """
    # Define as duplas
    duplas = {
        "Dupla_1": Dupla("Dupla_1", ["J1", "J3"]),
        "Dupla_2": Dupla("Dupla_2", ["J2", "J4"])
    }

    pontuacao_por_jogador = {nome: 0 for dupla in duplas.values() for nome in dupla.jogadores}

    rodadas = []
    jogador_inicial_nome = None  # definido dinamicamente a cada rodada

    # Loop principal
    while max(dupla.pontuacao for dupla in duplas.values()) < pontos_para_vencer:
        jogadores = distribuir_jogadores(estrategias)  # retorna List[Jogador]
        rodada = simular_rodada(
            jogadores,
            jogador_inicial_nome=jogador_inicial_nome,
            duplas=duplas,
            pontos_para_vencer=pontos_para_vencer,
        )

        if "erro" in rodada:
            continue  # refaz a rodada

        rodadas.append(rodada)
        final = rodada["final"]
        jogador_vencedor = final["vencedor_rodada"]
        pontos = final["pontuacao_rodada"]

        if jogador_vencedor:
            jogador_inicial_nome = jogador_vencedor  # define início da próxima rodada

            if duplas["Dupla_1"].contem_jogador(jogador_vencedor):
                duplas["Dupla_1"].adicionar_pontos(pontos)
            else:
                duplas["Dupla_2"].adicionar_pontos(pontos)

            pontuacao_por_jogador[jogador_vencedor] += pontos

        # Permite que estratégias aprendizes atualizem seus parâmetros
        for j in jogadores:
            estrategia = getattr(j, "estrategia", None)
            if hasattr(estrategia, "notificar_resultado"):
                estrategia.notificar_resultado(j.nome, jogador_vencedor)

    vencedor_partida = max(duplas.items(), key=lambda item: item[1].pontuacao)[0]

    return {
        "duplas": {nome: dupla.pontuacao for nome, dupla in duplas.items()},
        "pontuacao_por_jogador": pontuacao_por_jogador,
        "rodadas": rodadas,
        "vencedor_partida": vencedor_partida
    }


def salvar_resultado_em_csv(
    id_partida: str,
    resultado: dict,
    writer_partidas: csv.writer,
    writer_rodadas: csv.writer,
    writer_jogadas: csv.writer,
) -> None:
    """Registra o resultado de ``simular_partida`` em arquivos CSV já abertos."""

    writer_partidas.writerow(
        [
            id_partida,
            resultado["vencedor_partida"],
            resultado["pontuacao_por_jogador"]["J1"],
            resultado["pontuacao_por_jogador"]["J2"],
            resultado["pontuacao_por_jogador"]["J3"],
            resultado["pontuacao_por_jogador"]["J4"],
        ]
    )

    for id_rodada, rodada in enumerate(resultado["rodadas"]):
        final = rodada["final"]
        writer_rodadas.writerow(
            [
                id_partida,
                id_rodada,
                rodada["estados"][0]["jogador"],
                final["tipo_batida"],
                final["motivo_fim"],
                final["vencedor_rodada"],
                final["pontuacao_rodada"],
                resultado["pontuacao_por_jogador"]["J1"],
                resultado["pontuacao_por_jogador"]["J2"],
                resultado["pontuacao_por_jogador"]["J3"],
                resultado["pontuacao_por_jogador"]["J4"],
            ]
        )

        for estado in rodada["estados"]:
            if estado["tipo"] in ["jogada", "batida"] and estado["peca"]:
                x, y = estado["peca"]
            else:
                x, y = "", ""
            writer_jogadas.writerow(
                [
                    id_partida,
                    id_rodada,
                    estado["ordem_jogada"],
                    estado["jogador"],
                    estado["tipo"],
                    x,
                    y,
                    estado.get("lado", ""),
                ]
            )

pontuacao_jogadores = {"J1": 0, "J2": 0, "J3": 0, "J4": 0}

# NOVA FUNÇÃO PARA SALVAR O HISTÓRICO COMPLETO DAS PARTIDAS

def simular_varias_partidas_em_csv(n=10, pasta_destino="historico_csv"):
    os.makedirs(pasta_destino, exist_ok=True)
    
    partidas_csv = open(os.path.join(pasta_destino, "partidas.csv"), mode="w", newline="")
    rodadas_csv = open(os.path.join(pasta_destino, "rodadas.csv"), mode="w", newline="")
    jogadas_csv = open(os.path.join(pasta_destino, "jogadas.csv"), mode="w", newline="")

    writer_partidas = csv.writer(partidas_csv)
    writer_rodadas = csv.writer(rodadas_csv)
    writer_jogadas = csv.writer(jogadas_csv)

    writer_partidas.writerow(["id_partida", "vencedor_partida", "pontuacao_J1", "pontuacao_J2", "pontuacao_J3", "pontuacao_J4"])
    writer_rodadas.writerow(["id_partida", "id_rodada", "inicio_rodada", "tipo_batida", "motivo_fim", "vencedor_rodada", "pontuacao_rodada", "pontuacao_J1", "pontuacao_J2", "pontuacao_J3", "pontuacao_J4"])
    writer_jogadas.writerow(["id_partida", "id_rodada", "ordem_jogada", "jogador", "tipo", "peca_x", "peca_y", "lado"])


    for _ in range(n):
        id_partida = str(uuid.uuid4())
        resultado = simular_partida(pontuacao_por_jogador=pontuacao_jogadores)
        salvar_resultado_em_csv(
            id_partida,
            resultado,
            writer_partidas,
            writer_rodadas,
            writer_jogadas,
        )

    partidas_csv.close()
    rodadas_csv.close()
    jogadas_csv.close()

    
    print("\nExportação concluída. Arquivos salvos em:", pasta_destino)
    

if __name__ == "__main__":
    
    inicio = time.time()
    simular_varias_partidas_em_csv(n=10)
    fim = time.time()
    print(f"Tempo total de execução: {fim - inicio:.2f} segundos")

File: motor_de_jogo.js
function gerarPecas() {
  return Array.from({ length: 7 }, (_, i) => 
    Array.from({ length: 7 - i }, (_, j) => [i, j + i])
  ).flat();
}

function distribuirPecas() {
  const pecas = gerarPecas();
  for (let i = pecas.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [pecas[i], pecas[j]] = [pecas[j], pecas[i]];
  }
  
  const maos = {
    J1: pecas.slice(0, 6).sort((a, b) => a[0] - b[0] || a[1] - b[1]),
    J2: pecas.slice(6, 12).sort((a, b) => a[0] - b[0] || a[1] - b[1]),
    J3: pecas.slice(12, 18).sort((a, b) => a[0] - b[0] || a[1] - b[1]),
    J4: pecas.slice(18, 24).sort((a, b) => a[0] - b[0] || a[1] - b[1])
  };
  
  const pecasFora = pecas.slice(24);
  return [maos, pecasFora];
}

function jogadorComMaiorDuplo(maos) {
  let maiorDuplo = -1;
  let jogadorInicial = null;
  
  for (const [jogador, mao] of Object.entries(maos)) {
    for (const peca of mao) {
      if (peca[0] === peca[1] && peca[0] > maiorDuplo) {
        maiorDuplo = peca[0];
        jogadorInicial = jogador;
      }
    }
  }
  
  return [jogadorInicial, maiorDuplo];
}

function proximoJogador(jogadorAtual) {
  const ordem = ["J1", "J2", "J3", "J4"];
  const idx = ordem.indexOf(jogadorAtual);
  return ordem[(idx + 1) % 4];
}

function jogadasValidas(mao, pontas) {
  if (pontas[0] === null) return mao;
  return mao.filter(peca => 
    peca[0] === pontas[0] || peca[1] === pontas[0] || 
    peca[0] === pontas[1] || peca[1] === pontas[1]
  );
}

function jogarPeca(tabuleiro, pontas, peca) {
  if (!tabuleiro.length) {
    tabuleiro.push(peca);
    pontas[0] = peca[0];
    pontas[1] = peca[1];
    return "inicial";
  }
  
  if (peca[0] === pontas[0]) {
    tabuleiro.unshift([peca[1], peca[0]]);
    pontas[0] = peca[1];
    return "esquerda";
  } else if (peca[1] === pontas[0]) {
    tabuleiro.unshift(peca);
    pontas[0] = peca[0];
    return "esquerda";
  } else if (peca[0] === pontas[1]) {
    tabuleiro.push(peca);
    pontas[1] = peca[1];
    return "direita";
  } else if (peca[1] === pontas[1]) {
    tabuleiro.push([peca[1], peca[0]]);
    pontas[1] = peca[0];
    return "direita";
  }
}

function tipoDeBatida(pecaFinal, jogador, pontas) {
  const ehDuplo = pecaFinal[0] === pecaFinal[1];
  const encaixaEsquerda = pecaFinal[0] === pontas[0] || pecaFinal[1] === pontas[0];
  const encaixaDireita = pecaFinal[0] === pontas[1] || pecaFinal[1] === pontas[1];
  const encaixaAmbas = encaixaEsquerda && encaixaDireita;

  if (ehDuplo && encaixaAmbas) return "cruzada";
  if (!ehDuplo && encaixaAmbas && pontas[0] !== pontas[1]) return "la_e_lo";
  if (ehDuplo) return "carroca";
  if (encaixaDireita || encaixaEsquerda) return "simples";
}

function calcularPontuacaoBatida(tipoBatida) {
  const pontos = {
    simples: 1,
    carroca: 2,
    la_e_lo: 3,
    cruzada: 4
  };
  return pontos[tipoBatida] || 0;
}

function calcularPontuacaoTravamento(maos) {
  const somasMaos = Object.fromEntries(
    Object.entries(maos).map(([j, mao]) => [
      j,
      mao.reduce((sum, p) => sum + p[0] + p[1], 0)
    ])
  );
  
  const menorSoma = Math.min(...Object.values(somasMaos));
  const vencedores = Object.entries(somasMaos)
    .filter(([_, soma]) => soma === menorSoma)
    .map(([j]) => j);
  
  return vencedores.length > 1 ? [null, 0] : [vencedores[0], 1];
}

let placar = {
  "Dupla_1": 0,
  "Dupla_2": 0
};

let historicoRodadas = [];

function simularRodada() {
  const [maos, _] = distribuirPecas();
  const tabuleiro = [];
  const pontas = [null, null];
  const historico = [];
  const estados = [];
  let ordemJogada = 1;
  let [jogador, maiorDuplo] = jogadorComMaiorDuplo(maos);

  if (!jogador) {
    return { erro: "Nenhum duplo encontrado" };
  }

  let passesConsecutivos = 0;
  let vencedorRodada = null;
  let motivoFim = null;
  let tipoBatida = null;
  let pontuacaoRodada = 0;

  while (true) {
    const mao = maos[jogador];
    const jogadasDisponiveis = jogadasValidas(mao, pontas);
    const jogadas = [...jogadasDisponiveis];

    if (jogadas.length) {
      if (ordemJogada === 1) {
        const peca = [maiorDuplo, maiorDuplo];
        mao.splice(mao.findIndex(p => p[0] === maiorDuplo && p[1] === maiorDuplo), 1);
        const lado = jogarPeca(tabuleiro, pontas, peca);
        const tipo = mao.length === 0 ? "batida" : "jogada";
        historico.push({
          ordem: ordemJogada,
          jogador,
          tipo,
          peca,
          lado
        });
      } else {
        const peca = jogadas[0];
        mao.splice(mao.findIndex(p => p[0] === peca[0] && p[1] === peca[1]), 1);
        const lado = jogarPeca(tabuleiro, pontas, peca);
        const tipo = mao.length === 0 ? "batida" : "jogada";
        historico.push({
          ordem: ordemJogada,
          jogador,
          tipo,
          peca,
          lado
        });
        passesConsecutivos = 0;
        if (mao.length === 0) {
          tipoBatida = tipoDeBatida(peca, jogador, pontas);
          pontuacaoRodada = calcularPontuacaoBatida(tipoBatida);
          vencedorRodada = jogador;
          motivoFim = "batida";
        }
      }
    } else {
      historico.push({
        ordem: ordemJogada,
        jogador,
        tipo: "passe"
      });
      passesConsecutivos++;
      if (passesConsecutivos === 4) {
        [vencedorRodada, pontuacaoRodada] = calcularPontuacaoTravamento(maos);
        motivoFim = "travamento";
        tipoBatida = "travamento";
      }
    }

    estados.push({
      ordem_jogada: ordemJogada,
      jogador,
      tipo: historico[historico.length - 1].tipo,
      peca: historico[historico.length - 1].peca,
      lado: historico[historico.length - 1].lado,
      tabuleiro: [...tabuleiro],
      maos: JSON.parse(JSON.stringify(maos)),
      tipo_batida: tipoBatida,
      motivo_fim: motivoFim,
      vencedor_rodada: vencedorRodada,
      jogadas: [...historico],
      jogadas_disponiveis: jogadasDisponiveis
    });

    if (motivoFim) {
      if (vencedorRodada) {
        const dupla = ["J1", "J3"].includes(vencedorRodada) ? "Dupla_1" : "Dupla_2";
        placar[dupla] += pontuacaoRodada;
        historicoRodadas.push({
          vencedor: vencedorRodada,
          tipoBatida,
          pontos: pontuacaoRodada,
          dupla
        });
      }
      break;
    }

    ordemJogada++;
    jogador = proximoJogador(jogador);
  }

  return {
    estados,
    final: {
      tipo_batida: tipoBatida,
      motivo_fim: motivoFim,
      vencedor_rodada: vencedorRodada,
      pontuacao_rodada: pontuacaoRodada
    },
    placar,
    historicoRodadas
  };
}

module.exports = { simularRodada };